### PR TITLE
Fmi avi messageconverter tac 107

### LINKED ISSUE #107: Separate generic message parsing from generic bulletin parsing
Declare `ConversionSpecification`s for converting a single message as a TAC string to `GenericAviationWeatherMessage` and provide corresponding parser. Refactor current generic bulletin parser to use the new message parser.

The message `MessageType` is detected automatically. In case auto detection fails, two fallback methods based on conversion hints are used to infer message type, in following order:

1. `ConversionHints.KEY_MESSAGE_TYPE`: use the value.
1. `ConversionHints.KEY_BULLETING_HEADING` (does not yet exist, create it) containing a `BulletinHeading` object as value: use return value of `getExpectedContainedMessageType()`.

### PR DESCRIPTION
**Pull request by patrick-alaspaa**
closes #107

### CODE REVIEW

File: src/main/java/fi/fmi/avi/converter/tac/bulletin/GenericMeteorologicalBulletinParser.java [hunk 1]
```diff
@@ -88,158 +79,36 @@ public void setTACLexer(final AviMessageLexer lexer) {
...
             //the bulletinID could be used to provide full (or more complete message time info:
-            /*
             try {
                 bulletinMetadata = GTSExchangeFileInfo.Builder.from(bulletinID).build();
             } catch (IllegalArgumentException iae) {
-                result.addIssue(new ConversionIssue(ConversionIssue.Severity.WARNING, ConversionIssue.Type.SYNTAX, "Could not parse bulletin metadata "
-                        + "from bulletinID '" + bulletinID + "'"));
+                result.addIssue(new ConversionIssue(ConversionIssue.Severity.WARNING, ConversionIssue.Type.SYNTAX,
+                        "Could not parse bulletin metadata " + "from bulletinID '" + bulletinID + "'"));
             }
```
Comment (petringo) [inline src/main/java/fi/fmi/avi/converter/tac/bulletin/GenericMeteorologicalBulletinParser.java:92]: Is there a reason for this being uncommented? Right now I don't remember the actual reason for this being commented out, but I have a hunch it lead to some problems.

File: src/main/java/fi/fmi/avi/converter/tac/conf/TACConverter.java [hunk 2]
```diff
@@ -115,6 +116,12 @@
     public static final ConversionSpecification<GenericMeteorologicalBulletin, String> GENERIC_BULLETIN_POJO_TO_TAC = new ConversionSpecification<>(
             GenericMeteorologicalBulletin.class, String.class, null, "WMO GTS bulletin");
 
+    /**
+     *
+     */
+    public static final ConversionSpecification<String, GenericAviationWeatherMessage> TAC_TO_GENERIC_AVIATION_WEATHER_MESSAGE_POJO =
+            new ConversionSpecification<>(String.class, GenericAviationWeatherMessage.class, "TAC string", null);
```
Comment (petringo) [inline src/main/java/fi/fmi/avi/converter/tac/conf/TACConverter.java:123]: I would use `ICAO Annex 3 TAC` as output specifier, just like in other message conversions.

File: src/test/java/fi/fmi/avi/converter/tac/bulletin/GenericAviationWeatherMessageParserTest.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        GenericAviationWeatherMessage msg = result.getConvertedMessage().get();
+
+        assertEquals(tacMessage, msg.getOriginalMessage());
+        assertEquals(MessageType.TAF, msg.getMessageType().get());
+        assertEquals(GenericAviationWeatherMessage.Format.TAC, msg.getMessageFormat());
+        assertEquals("--30T11:03Z",
+                msg.getIssueTime().map(PartialOrCompleteTimeInstant::getPartialTime).map(Optional::get).map(PartialDateTime::toString).orElse(null));
+        assertEquals("--30T12:Z", msg.getValidityTime()
+                .map(PartialOrCompleteTimePeriod::getStartTime)
+                .map(Optional::get)
```
Comment (petringo) [inline src/test/java/fi/fmi/avi/converter/tac/bulletin/GenericAviationWeatherMessageParserTest.java:59]: Hint: `map(PartialOrCompleteTimeInstant::getStartTime).map(Optional::get)` and similar cases can be written shorter and clearer `flatMap(PartialOrCompleteTimeInstant::getStartTime)`. I would suggest to do a global search for `map(Optional::get)`.

File: src/test/java/fi/fmi/avi/converter/tac/TACTestConfiguration.java [hunk 4]
```diff
@@ -75,6 +76,10 @@
     @Autowired
     private AviMessageSpecificConverter<SpaceWeatherBulletin, String> swxBulletinTACSerializer;
 
+    @Autowired
+    @Qualifier("GenericAviationWeatherMessageTACParser")
+    private AviMessageSpecificConverter<String, GenericAviationWeatherMessage> TACGenericAviationWeatherMessageParser;
```
Comment (petringo) [inline src/test/java/fi/fmi/avi/converter/tac/TACTestConfiguration.java:81]: To retain consistent naming I would rename this to `genericAviationWeatherMessageTACParser`.

**Review by petringo — CHANGES_REQUESTED** [review]

File: src/main/java/fi/fmi/avi/converter/tac/bulletin/GenericMeteorologicalBulletinParser.java (hunk 1 above)
Comment (patrick-alaspaa) [reply src/main/java/fi/fmi/avi/converter/tac/bulletin/GenericMeteorologicalBulletinParser.java:93]: Seems that I accidentally uncommented this section of code.

File: src/test/java/fi/fmi/avi/converter/tac/TACTestConfiguration.java (hunk 4 above)
Comment (petringo) [reply src/test/java/fi/fmi/avi/converter/tac/TACTestConfiguration.java]: In addition, I think the qualifier is redundant, and the bean name should start with lower case letter. I can change this upon merge.

**Review by petringo — APPROVED** [review]
I think this is ready for merging now. I'll just do [this little change](https://github.com/fmidev/fmi-avi-messageconverter-tac/pull/110#discussion_r666077123) upon merge.

**Pull request merged**